lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -3223,15 +3259,44 @@ private long prepareCommitInternal() throws IOException {
               // sneak into the commit point:
               toCommit = segmentInfos.clone();
 
+              if (anyChanges) {
+                mergeAwaitLatchRef = new AtomicReference<>();
```

Comment:
could we wait to create this until we create its CountdownLatch and simplify the null checks below?

Reply:
This is a little bit of hackery to share state between this thread and the threads that do the merges. 

We initialize the ref here, pass it to `waitForMergeOnCommitPolicy` on the next line to make sure it gets shared with any computed `OneMerge`s. Then, before we release the IW lock (so we're guaranteed that those `OneMerge`s haven't run yet), we populate the ref with the `CountdownLatch` (once we know what we're counting down).

That said, I think I could simplify things a lot by not using `OneMergeWrappingMergePolicy`, but rather decorating the returned `OneMerge`s (if applicable) directly. I'm going to take a stab at that approach in my next commit.